### PR TITLE
Add specialization for sum of large tuples.

### LINKED ISSUE #35350: Sum of large-ish tuple of arrays leads to StackOverFlowError
## Reproducing
Calculating the sum of a sufficiently large NTuple containing numeric arrays leads to a StackOverFlowError. The breaking point seems to be somewhere around ~1,000 items in the tuple.
```
julia> datavals = ntuple(i->[3.0, 4.0], 1600);
julia> sum(datavals)
Internal error: encountered unexpected error in runtime:
StackOverflowError()
...
```
The stacktrace goes on for quite some time, with the repeating bit:
```
abstract_call_method at ./compiler/abstractinterpretation.jl:404                                                                 
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:101                                                             
abstract_call_known at ./compiler/abstractinterpretation.jl:873                                                                  
abstract_call at ./compiler/abstractinterpretation.jl:895                                                                        
abstract_apply at ./compiler/abstractinterpretation.jl:604                                                                       
abstract_call_known at ./compiler/abstractinterpretation.jl:673                                                                  
abstract_call at ./compiler/abstractinterpretation.jl:895                                                                        
abstract_call at ./compiler/abstractinterpretation.jl:880
abstract_eval at ./compiler/abstractinterpretation.jl:974
typeinf_local at ./compiler/abstractinterpretation.jl:1227
typeinf_nocycle at ./compiler/abstractinterpretation.jl:1283
typeinf at ./compiler/typeinfer.jl:12
```

## Version info
```
julia> versioninfo()
Julia Version 1.4.0
Commit b8e9a9ecc6 (2020-03-21 16:36 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD Ryzen Threadripper 1950X 16-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, znver1)
```

### PR DESCRIPTION
**Pull request by dmeklund**
The default definition of +(x...) would lead to a stack overflow
with sufficiently large tuples. Remove seemingly outdated comments
and add a new test.

Fixes #35350

The first TODO removed definitely seems irrelevant. The second pair are a bit more vague to me, I'm not sure what a "regular" sum or prod is supposed to refer to, but I assume whatever size promotion is expected will be provided by passing along the arguments to `+` or `*`. I tried to dig into the git history to see if I could find when the comment(s) became irrelevant but was unable to find anything.

The updated implementation is not fast for a tuple of arrays (~0.1 seconds to sum a 1600 element ntuple) but optimization would require further specialization for array arguments for a case that is presumably unusual.

### CODE REVIEW

File: base/tuple.jl [hunk 1]
```diff
@@ -389,15 +389,17 @@ reverse(t::Tuple) = revargs(t...)
 
 ## specialized reduction ##
 
-# TODO: these definitions cannot yet be combined, since +(x...)
-# where x might be any tuple matches too many methods.
-# TODO: this is inconsistent with the regular sum in cases where the arguments
```
Comment (JeffBezanson) [inline base/tuple.jl:394]: This comment (the second TODO) is still relevant --- it refers to e.g. `sum(Int8[1])` vs. `sum((Int8(1),))`.

Comment (dmeklund) [reply base/tuple.jl:394]: Thanks. I'll add back and clarify the TODO.

File: base/tuple.jl [hunk 2]
```diff
@@ -389,15 +389,17 @@ reverse(t::Tuple) = revargs(t...)
...
-# require size promotion to system size.
 sum(x::Tuple{Any, Vararg{Any}}) = +(x...)
+# avoid stack overflow with large tuples
+function sum(x::Any16)
```
Comment (stevengj) [inline base/tuple.jl:394]: Or just `sum(x::Any16) = foldl(+, x)`?

**Comment by laborg** [comment]
@dmeklund in recent Julia versions sum of large tuples is decently fast:
```
julia> using BenchmarkTools

julia> @btime sum(ntuple(i->[1.0, 1.0], 1600000))
  306.607 ms (3200003 allocations: 280.76 MiB)
2-element Vector{Float64}:
 1.6e6
 1.6e6
``` 
Mind to close or change this PR?

**Comment by dmeklund** [comment]
@laborg Thanks for the heads up, and happy to hear. Closing.

**Pull request closed without merging**